assets/src/edit-story/components/library/panes/media/common/paginatedMediaGallery.js

```diff
@@ -0,0 +1,145 @@
+  providerType,
+  resources,
+  isMediaLoading,
+  isMediaLoaded,
+  hasMore,
+  onInsert,
+  setNextPage,
+}) {
+  // TODO(#1698): Ensure scrollbars auto-disappear in MacOS.
+  // State and callback ref necessary to recalculate the padding of the list
+  // given the scrollbar width.
+  const [scrollbarWidth, setScrollbarWidth] = useState(0);
+  const refContainer = useRef();
+  const refCallbackContainer = (element) => {
+    refContainer.current = element;
+    if (!element) {
+      return;
+    }
+    setScrollbarWidth(element.offsetWidth - element.clientWidth);
+  };
+
+  // Recalculates padding so that it stays centered.
+  // As of May 2020 this cannot be achieved without js (as the scrollbar-gutter
+  // prop is not yet ready).
+  useLayoutEffect(() => {
+    if (!scrollbarWidth) {
+      return;
+    }
+    const currentPaddingLeft = parseFloat(
+      window
+        .getComputedStyle(refContainer.current, null)
+        .getPropertyValue('padding-left')
+    );
+    refContainer.current.style['padding-right'] =
+      currentPaddingLeft - scrollbarWidth + 'px';
+  }, [scrollbarWidth, refContainer]);
+
+  const refContainerFooter = useRef();
+  useIntersectionEffect(
+    refContainerFooter,
+    {
+      root: refContainer,
+      // This rootMargin is added so that we load an extra page when the
+      // "loading" footer is "close" to the bottom of the container, even if
+      // it's not yet visible.
+      rootMargin: `0px 0px ${ROOT_MARGIN}px 0px`,
+    },
+    (entry) => {
+      if (
+        !isMediaLoaded ||
+        isMediaLoading ||
+        !hasMore ||
+        !entry.isIntersecting
+      ) {
+        return;
+      }
+      setNextPage();
+    },
+    [hasMore, isMediaLoading, isMediaLoaded, setNextPage]
+  );
+
+  const mediaGallery =
+    isMediaLoaded && resources.length === 0 ? (
+      <MediaGalleryMessage>
+        {__('No media found', 'web-stories')}
+      </MediaGalleryMessage>
+    ) : (
+      <>
+        <div style={{ marginBottom: 15 }}>
+          <MediaGallery
+            providerType={providerType}
+            resources={resources}
+            onInsert={onInsert}
+          />
+        </div>
+        {hasMore && (
+          <MediaGalleryLoadingPill ref={refContainerFooter}>
+            {__('Loading…', 'web-stories')}
+          </MediaGalleryLoadingPill>
+        )}
+      </>
+    );
+
+  return (
+    <MediaGalleryContainer
+      data-testid="media-gallery-container"
+      ref={refCallbackContainer}
+    >
+      {mediaGallery}
+    </MediaGalleryContainer>
+  );
+}
+
+PaginatedMediaGallery.propTypes = {
+  providerType: PropTypes.string.isRequired,
+  resources: PropTypes.arrayOf(PropTypes.object).isRequired,
+  isMediaLoading: PropTypes.bool.isRequired,
+  isMediaLoaded: PropTypes.bool.isRequired,
+  hasMore: PropTypes.bool.isRequired,
+  onInsert: PropTypes.func.isRequired,
+  setNextPage: PropTypes.func.isRequired,
+};
+
+export default memo(PaginatedMediaGallery);
```

Comment:
what does applying memo here do?

Reply:
This marks the component as a 'pure component', which is good practice for components that have no children, though is not always strictly necessary. If none of the properties have changed using reference equality, then the component is not re-rendered.